ta/bonnie/bonnie.c

```diff
@@ -0,0 +1,406 @@
+/*
+ * Copyright (c) 2015, Linaro Limited
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <tee_api.h>
+#include <bonnie.h>
+#include <ta_bonnie.h>
+#include <tee_internal_api_extensions.h>
+
+#define DEFAULT_CHUNK_SIZE (1 << 12)
+#define DEFAULT_DATA_SIZE (DEFAULT_CHUNK_SIZE * 1023)
+
+#define SCRAMBLE(x) ((x & 0xff) ^ 0xaa)
+
+#define ASSERT_PARAM_TYPE(pt_in, pt_expect) \
+do { \
+	if ((pt_in) != (pt_expect)) \
+		return TEE_ERROR_BAD_PARAMETERS; \
+} while (0)
+
+
+static uint8_t filename[] = "BonnieTestFile";
+
+static void fill_buffer(uint8_t *buf, size_t size)
+{
+	size_t i;
+
+	if (!buf)
+		return;
+
+	for (i = 0; i < size; i++)
+		buf[i] = SCRAMBLE(i);
+}
+
+static TEE_Result verify_buffer(uint8_t *buf, size_t size)
+{
+	size_t i;
+
+	if (!buf)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	for (i = 0; i < size; i++) {
+		uint8_t expect_data = SCRAMBLE(i);
+
+		if (expect_data != buf[i]) {
+			return TEE_ERROR_CORRUPT_OBJECT;
+		}
+	}
+
+	return TEE_SUCCESS;
+}
+
+static inline uint32_t tee_time_to_ms(TEE_Time t)
+{
+	return t.seconds * 1000 + t.millis;
+}
+
+static inline uint32_t get_delta_time_in_ms(TEE_Time start, TEE_Time stop)
+{
+	return tee_time_to_ms(stop) -tee_time_to_ms(start);
+}
+
+static TEE_Result test_write(TEE_ObjectHandle object, size_t data_size,
+		uint8_t *chunk_buf, size_t chunk_size,
+		uint32_t *spent_time_in_ms)
+{
+	TEE_Time start_time, stop_time;
+	size_t tmp_data_size = data_size;
+	TEE_Result res = TEE_SUCCESS;
+
+	/* Seek position to the beginning of the second block */
+	res = TEE_SeekObjectData(object, 4080, TEE_DATA_SEEK_SET);
```

Comment:
4080 is a `hard-coded` value that may depend on the implementation (same below).

Would be good to have it as an argument (block start, middle of the block, end of block) to check how it behaves. I would expect "write" does not depend too much on the block location, whereas read is faster at the beginning of the block.